service-base/src/main/java/org/eclipse/hono/service/tenant/TenantHttpEndpoint.java

```diff
@@ -68,155 +69,104 @@ public String getName() {
-                        case HttpURLConnection.HTTP_CREATED:
-                            response.putHeader(
-                                    HttpHeaders.LOCATION,
-                                    String.format("/%s/%s", TenantConstants.TENANT_ENDPOINT, tenantId));
-                        default:
-                            if (status >= 400) {
-                                setResponseBody(tenantResult, response);
-                            }
-                            response.end();
-                    }
-                });
-            }
+        if (tenantId == null) {
+            ctx.response().setStatusMessage(String.format("'%s' param is required", FIELD_TENANT_ID));
+            ctx.fail(HttpURLConnection.HTTP_BAD_REQUEST);
+        } else if (!(tenantId instanceof String)) {
+            ctx.response().setStatusMessage(String.format("'%s' must be a string", FIELD_TENANT_ID));
+            ctx.fail(HttpURLConnection.HTTP_BAD_REQUEST);
         }
+        ctx.next();
     }
 
-    private void doUpdateTenantJson(final RoutingContext ctx) {
+    private void doTenantHttpRequest(final RoutingContext ctx, final TenantConstants.Action action, final Predicate<Integer> sendResponseForStatus) {
 
-        try {
-            JsonObject payload = null;
-            if (ctx.getBody().length() > 0) {
-                payload = ctx.getBodyAsJson();
-            }
-            updateTenant(payload, ctx);
-        } catch (final DecodeException e) {
-            HttpUtils.badRequest(ctx, "body does not contain a valid JSON object");
-        }
-    }
+        final JsonObject payload = ctx.get(KEY_REQUEST_BODY);
 
-    private void updateTenant(final JsonObject payload, final RoutingContext ctx) {
+        final String tenantId = Optional.ofNullable(getTenantParam(ctx)).orElse(getTenantParamFromPayload(payload));
+        logger.debug("http request [{}] for tenant [tenant: {}]", action, tenantId);
 
-        final String tenantId = getTenantParam(ctx);
-        logger.debug("updating tenant [tenant: {}, payload: {}]", tenantId, payload);
-        if (payload != null) {
-            payload.remove(FIELD_TENANT_ID);
-        }
         final HttpServerResponse response = ctx.response();
-        final JsonObject requestMsg = TenantConstants.getServiceRequestAsJson(Action.ACTION_UPDATE.toString(),
-                tenantId, payload);
+        final JsonObject requestMsg = TenantConstants.getServiceRequestAsJson(action.toString(), tenantId, payload);
 
-        doTenantAction(ctx, requestMsg, (status, tenantResult) -> {
-                response.setStatusCode(status);
-                if (status >=400) {
+        sendTenantAction(ctx, requestMsg, (status, tenantResult) -> {
+            response.setStatusCode(status);
+            if (status >=400) {
+                setResponseBody(tenantResult, response);
+            } else if (sendResponseForStatus != null) {
+                if (sendResponseForStatus.test(status)) {
```

Comment:
why not let another handler process the response? we can provide (and register) default handlers for the CRUD responses which can be overridden by sub-classes if appropriate